sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts

```diff
@@ -54,4 +57,56 @@ export class ApplicationService extends RecordDataModelService<Application> {
 
     return updateResult.affected;
   }
+
+  /**
+   * Finds all applications with some pending student assessment to be
+   * processed by the workflow, ignoring applications that already have
+   * some workflows in progress or were already queued for execution.
+   * All pending student assessments will be returned ordered by its creation
+   * date to allow the select of the next one to be executed (usually only one
+   * record would be expected).
+   * @returns applications with pending assessments to be executed.
+   */
+  async getApplicationsToStartAssessments(): Promise<Application[]> {
+    // Sub query to validate if an application has assessment already being
+    // processed by the workflow.
+    const inProgressStatusesExistsQuery = this.studentAssessmentRepo
+      .createQueryBuilder("studentAssessment")
+      .select("1")
+      .where("studentAssessment.application.id = application.id")
+      .andWhere(
+        "studentAssessment.studentAssessmentStatus IN (:...inProgressStatuses)",
+      )
+      .getQuery();
+    return (
+      this.applicationRepo
```

Comment:
Just wondering, this query could have come from student assessments instead of applications right? so that it will just return the assesments(assessment id) which needs to be processed.

Reply:
True. It was originally designed to allow the later update of the application + assessment together but it did not work as expected and I ended up doing the individual updates in a transaction.
Right now the only benefit is that it makes it pretty easy to have the assessments grouped to allow the processing and also pick the next assessment to be processed if multiple exists under the same application.